File: cs/usage-maven-plugin/src/main/java/com/usalko/maven/plugin/usage/asm/UsageClassVisitor.java
package com.usalko.maven.plugin.usage.asm;

/*
 * Copyright 2001-2005 The Apache Software Foundation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import org.objectweb.asm.AnnotationVisitor;
import org.objectweb.asm.Attribute;
import org.objectweb.asm.ClassVisitor;
import org.objectweb.asm.FieldVisitor;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.TypePath;

public class UsageClassVisitor extends ClassVisitor {

    private final VisitorsContext visitorsContext;

    public UsageClassVisitor(VisitorsContext context) {
        super( Opcodes.ASM7 );
        this.visitorsContext = context;
    }

    @Override
    public void visit(int version, int access, String name, String signature, String superName,
            String[] interfaces) {
        super.visit(version, access, name, signature, superName, interfaces);
        visitorsContext.setCurrentClassName(name);
    }

    @Override
    public AnnotationVisitor visitAnnotation(String descriptor, boolean visible) {
        return new UsageAnnotationVisitor(visitorsContext);
    }

    @Override
    public AnnotationVisitor visitTypeAnnotation(int typeRef, TypePath typePath, String descriptor,
            boolean visible) {
        return new UsageAnnotationVisitor(visitorsContext);
    }

    @Override
    public void visitAttribute(Attribute attribute) {
        super.visitAttribute(attribute);
        visitorsContext.acceptClass(attribute.type);
    }

    @Override
    public void visitNestMember(String nestMember) {
        super.visitNestMember(nestMember);
        visitorsContext.acceptClass(nestMember);
    }

    @Override
    public void visitInnerClass(String name, String outerName, String innerName, int access) {
        super.visitInnerClass(name, outerName, innerName, access);
        visitorsContext.acceptClass(innerName);
    }

    @Override
    public FieldVisitor visitField(int access, String name, String descriptor, String signature,
            Object value) {
        return new UsageFieldVisitor(visitorsContext);
    }

    @Override
    public MethodVisitor visitMethod(int access, String name, String descriptor, String signature,
            String[] exceptions) {
        return new UsageMethodVisitor(visitorsContext);
    }

}

File: cs/gradle-usage-plugin/src/main/kotlin/com/usalko/gradle/plugin/asm/UsageClassVisitor.kt
package com.usalko.gradle.plugin.asm

/*
 * Copyright 2001-2005 The Apache Software Foundation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import org.objectweb.asm.*

class UsageClassVisitor(private val visitorsContext: VisitorsContext) : ClassVisitor(Opcodes.ASM7) {

    override fun visit(version: Int, access: Int, name: String, signature: String?,
                       superName: String?, interfaces: Array<String>) {
        super.visit(version, access, name, signature, superName, interfaces)
        visitorsContext.setCurrentClassName(name)
    }

    override fun visitAnnotation(descriptor: String, visible: Boolean): AnnotationVisitor {
        return UsageAnnotationVisitor(visitorsContext)
    }

    override fun visitTypeAnnotation(typeRef: Int, typePath: TypePath, descriptor: String,
                                     visible: Boolean): AnnotationVisitor {
        return UsageAnnotationVisitor(visitorsContext)
    }

    override fun visitAttribute(attribute: Attribute) {
        super.visitAttribute(attribute)
        visitorsContext.acceptClass(attribute.type)
    }

    override fun visitNestMember(nestMember: String) {
        super.visitNestMember(nestMember)
        visitorsContext.acceptClass(nestMember)
    }

    override fun visitInnerClass(name: String, outerName: String?, innerName: String?, access: Int) {
        super.visitInnerClass(name, outerName, innerName, access)
        visitorsContext.acceptClass(innerName)
    }

    override fun visitField(access: Int, name: String, descriptor: String, signature: String?,
                            value: Any?): FieldVisitor {
        return UsageFieldVisitor(visitorsContext)
    }

    override fun visitMethod(access: Int, name: String, descriptor: String, signature: String?,
                             exceptions: Array<String>?): MethodVisitor {
        return UsageMethodVisitor(visitorsContext)
    }

}
